### PR TITLE
Collision code refactor

### PR DESCRIPTION
**Pull request by mikadou**
Refactor, cleanup and documentation of some collision related code.
Does not generate any performance regression for me 

| scenario | PR              | master |
| ----------- | -------------- | --------- |
| Container test | 9.5 FPS | 9.5 FPS |
| 12 cars | 6.7 FPS | 6.6 FPS |
| 6 cars | 12.8 FPS | 12.3 FPS |
| 5 cars | 14.4 FPS | 14.2 FPS |
| 4 cars | 17.8 FPS | 17.8 FPS |
| 3 cars | 20.0 FPS | 20.1 FPS |

### CODE REVIEW

**Comment by ulteq** [comment]
Is there a special reason why the PointColDetector update routine is not called within the collision functions?

* https://github.com/mikadou/rigs-of-rods/blob/collision-refactor/source/main/physics/Beam.cpp#L1547
* https://github.com/mikadou/rigs-of-rods/blob/collision-refactor/source/main/physics/BeamFactory.cpp#L1160

File: source/main/physics/collision/DynamicCollisions.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        const int free_collcab, int collcabs[], int cabs[],
+        collcab_rate_t inter_collcabrate[], node_t nodes[],
+        const float collrange, Beam **trucks,
+        const int numtrucks,
+        ground_model_t &submesh_ground_model)
+{
+    //If you change any of the below "ifs" concerning trucks then you should
+    //also consider changing the parallel "ifs" inside PointColDetector
+    //see "pointCD" above.
+    //Performance some times forces ugly architectural designs....
```
Comment (ulteq) [inline source/main/physics/collision/DynamicCollisions.cpp]: I think it is time to finally remove that comment.

File: source/main/physics/collision/CartesianToTriangleTransform.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    TriangleCoord operator() (const Vector &p) const
+    {
+        // lazy initialization of transformation matrix
+        if (!m_matrix) {
+            InitMatrix();
+        }
+
+        // apply transformation matrix and extract alpha, beta, gamma and perpendicular offset
+        const Vector result = *m_matrix * (p - m_triangle.c);
+        return {result[0], result[1], (1.0 - result[0] - result[1]), result[2]};
```
Comment (ulteq) [inline source/main/physics/collision/CartesianToTriangleTransform.h]: > warning: narrowing conversion of ‘(((-(double)result.Ogre::Vector3::operator[](0ul)) - ((double)result.Ogre::Vector3::operator[](1ul))) + 1.0e+0)’ from ‘double’ to ‘Real {aka float}’ inside { } [-Wnarrowing]
         return {result[0], result[1], (1.0 - result[0] - result[1]), result[2]};

**Comment by ulteq** [comment]
> Does not generate any performance regression for me

I did not find any differences in performance either. :+1:

File: source/main/physics/collision/CartesianToTriangleTransform.h (hunk 2 above)
Comment (mikadou) [inline source/main/physics/collision/CartesianToTriangleTransform.h]: Thanks, fixed 63f47a5

File: source/main/physics/collision/DynamicCollisions.cpp (hunk 1 above)
Comment (mikadou) [inline source/main/physics/collision/DynamicCollisions.cpp]: Agreed, a3adab0

**Comment by mikadou** [comment]
> Is there a special reason why the PointColDetector update routine is not called within the collision functions?

There might have been a reason, but I can't exactly remember. How much does it bother you?

**Comment by ulteq** [comment]
> How much does it bother you?

Not much, just leave it as is.

**Pull request merged**